File: YuXiangROS/Chap10/Advance_ws/src/learn_lifecyclenode_cpp/src/learn_lifecyclenode.cpp
#include <rclcpp/rclcpp.hpp>
#include <rclcpp_lifecycle/lifecycle_node.hpp>

using CallbackReturn = rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn;

class LearnLifeCycleNode : public rclcpp_lifecycle::LifecycleNode {
  private:
    double timer_period_;
    rclcpp::TimerBase::SharedPtr timer_;

  public:
    LearnLifeCycleNode() : rclcpp_lifecycle::LifecycleNode("lifecyclenode") {
        timer_period_ = 1.0;
        timer_ = nullptr;
        RCLCPP_INFO(get_logger(), "%s: has been created", get_name());
    }

    CallbackReturn on_configure(const rclcpp_lifecycle::State& state) override {
        (void)state;
        timer_period_ = 1.0;
        RCLCPP_INFO(get_logger(), "on_configure(): Setup timer_period");
        return rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn::SUCCESS;
    }

    CallbackReturn on_activate(const rclcpp_lifecycle::State& state) override {
        (void)state;
        timer_ = create_wall_timer(
            std::chrono::seconds(static_cast<int>(timer_period_)),
            std::bind(&LearnLifeCycleNode::timer_callback_, this)
        );
        RCLCPP_INFO(get_logger(), "on_activate(): Create timer");
        return rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn::SUCCESS;
    }

    CallbackReturn on_deactivate(const rclcpp_lifecycle::State& state) override {
        (void)state;
        timer_.reset();
        RCLCPP_INFO(get_logger(), "on_deactivate(): Destroy timer");
        return rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn::SUCCESS;
    }

    CallbackReturn on_shutdown(const rclcpp_lifecycle::State& state) override {
        (void)state;
        timer_.reset();
        RCLCPP_INFO(get_logger(), "on_shutdown()");
        return rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn::SUCCESS;
    }

  private:
    void timer_callback_() { RCLCPP_INFO(get_logger(), "Timer is outputting..."); }
};

int main(int argc, char* argv[]) {
    rclcpp::init(argc, argv);
    auto node = std::make_shared<LearnLifeCycleNode>();
    rclcpp::spin(node->get_node_base_interface());
    rclcpp::shutdown();
    return 0;
}

File: YuXiangROS/Chap10/Advance_ws/src/learn_lifecyclenode_py/learn_lifecyclenode_py/learn_lifecyclenode.py
import rclpy
from rclpy.lifecycle import LifecycleNode, TransitionCallbackReturn


class LearnLifeCycleNode(LifecycleNode):
    """
    设置生命状态
    ros2 lifecycle set /lifecyclenode [state]

    state:
    configure, activate, deactivate, shutdown
    """

    def __init__(self):
        super().__init__("lifecyclenode")
        self.timer_period = 0
        self.timer_ = None
        self.get_logger().info(f"{self.get_name()}: has been created")

    def timer_callback(self):
        self.get_logger().info("Timer is outputting...")

    def on_configure(self, state):
        self.timer_period = 1.0  # 设置定时器周期
        self.get_logger().info("on_configure(): Setup timer_period")
        return TransitionCallbackReturn.SUCCESS

    def on_activate(self, state):
        self.timer_ = self.create_timer(self.timer_period, self.timer_callback)
        self.get_logger().info("on_activate(): Create timer")
        return TransitionCallbackReturn.SUCCESS

    def on_deactivate(self, state):
        if self.timer_ is None:
            self.get_logger().warn("on_deactivate(): Timer has been destroyed!")
            return TransitionCallbackReturn.SUCCESS
        self.destroy_timer(self.timer_)  # 销毁定时器
        self.get_logger().info("on_deactivate(): Destroy timer")
        return TransitionCallbackReturn.SUCCESS

    def on_shutdown(self, state):
        # 定时器未销毁则销毁
        if self.timer_ is not None:
            self.destroy_timer(self.timer_)
        self.get_logger().info("on_shutdown()")
        return TransitionCallbackReturn.SUCCESS


def main():
    rclpy.init()
    node = LearnLifeCycleNode()
    rclpy.spin(node)
    rclpy.shutdown()
